File: src/main/java/mnm/mods/tabbychat/client/gui/Scrollbar.java
package mnm.mods.tabbychat.client.gui;

import mnm.mods.tabbychat.util.ILocation;
import mnm.mods.tabbychat.client.gui.component.GuiComponent;
import net.minecraft.client.gui.Gui;

public class Scrollbar extends GuiComponent {

    private ChatArea chat;

    public Scrollbar(ChatArea chat) {
        this.chat = chat;
    }

    @Override
    public void render(int mouseX, int mouseY, float parTicks) {
        if (mc.ingameGUI.getChatGUI().getChatOpen()) {
            int scroll = chat.getScrollPos();
            int max = chat.getLocation().getHeight();
            int lines = max / mc.fontRenderer.FONT_HEIGHT;
            int total = chat.getChat().size();
            if (total <= lines) {
                return;
            }
            total -= lines;
            int size = Math.max(max / 2 - total, 10);
            float perc = Math.abs((float) scroll / (float) total - 1) * Math.abs((float) size / (float) max - 1);
            int pos = (int) (perc * max);

            ILocation loc = getLocation();
            Gui.drawRect(loc.getXPos(), loc.getYPos() + pos, loc.getXPos() + 1, loc.getYPos() + pos + size, -1);
            super.render(mouseX, mouseY, parTicks);
        }
    }

}

File: src/main/kotlin/mnm/mods/tabbychat/client/gui/Scrollbar.kt
package mnm.mods.tabbychat.client.gui

import mnm.mods.tabbychat.util.mc
import net.minecraft.client.gui.AbstractGui
import net.minecraft.client.gui.IRenderable
import kotlin.math.absoluteValue
import kotlin.math.max

class Scrollbar(private val chat: ChatArea) : AbstractGui(), IRenderable {

    override fun render(x: Int, y: Int, parTicks: Float) {
        if (mc.ingameGUI.chatGUI.chatOpen) {
            val scroll = chat.scrollPos.toFloat()
            val max = chat.height.toFloat()
            val lines = max / mc.fontRenderer.FONT_HEIGHT
            var total = chat.chat.size.toFloat()
            if (total <= lines) {
                return
            }
            total -= lines
            val size = max(max / 2 - total, 10.toFloat())
            val perc = ((scroll / total - 1) * (size / max - 1)).absoluteValue
            val pos = (perc * max).toInt()

            fill(chat.xPos + chat.width,
                    chat.yPos + pos,
                    chat.xPos + chat.width + 1,
                    chat.yPos + pos + size.toInt(), -1)
        }
    }

}
